### DISCUSSION TITLE
Quadrature point output

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#25091] [Q&A General] [answered]
I'm interested in saving material (quadrature point) data for postprocessing. Is there a supported mesh output format capable of storing quadrature point data? If not, is there a way to modify a VPP (like MaterialVectorPostprocessor) to apply to all elements?

**Comment by GiudGiud** [comment]
No luck on Qp output on meshes. 

Modifying MaterialVPP to have the option to be either:
- block restricted
- on a list of elements (only thing we have now)
- on all elements

seems like a great thing for MOOSE

**Reply by hugary1995** [reply]
Some day moose should support point cloud output format, and then we can write all material properties easily :)

**Reply by jessecarterMOOSE** [reply]
I opened an issue (#25093) but might try to do this myself. Couple of questions along those lines:
1. Currently the list of element ids comes from the user and gets saved as a set and looped over. How can I default that to a list of all element ids?
2. How can I also save the qp coordinates along with the material property data?

**Reply by GiudGiud** [reply]
@hugary1995 we have the Positions system now so that could be the way to save the point cloud coordinates. 
We dont have the output yet but I agree it would be good. Something like "FunctionAtPositionsOutput" would cover so many needs

in fact this is where I m headed there:
https://github.com/idaholab/moose/pull/24412
but I m not quite there

@jessecarterMOOSE 
You can make this non-const and build it in the constructor (from the mesh element list)
For 2, follow the same patter, first index by element id, then index by quadrature point index

**Reply by jessecarterMOOSE** [reply]
> @jessecarterMOOSE You can make this non-const and build it in the constructor (from the mesh element list) For 2, follow the same patter, first index by element id, then index by quadrature point index

You make it sound so easy, but you may be overestimating my ability. Are there any examples of getting a list of element ids in the framework? And how do I get at qp coordinates from inside that VPP? I haven't developed VPP's before.

**Reply by GiudGiud** [reply]
If you use an ElementVPP (or a class deriving from ElementUO) you have the _q_point array that can be accessed at _qp to get quadrature points. 
Doing this may be easier actually than trying to modify the MaterialVPP...
You ll get block restriction etc already done well. 

It's not clear to me actually what s best. I have not edited VPPs for that purpose before. 

to get a list of ids use this:
getMesh() on the MooseMesh (_mesh in the class probably) to get the libmesh meshbase
then 
this routine to get a range of element on the blocks desired
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1MeshBase.html#ad3694fca532d45f4ad343ef66dfe2190

**Reply by jessecarterMOOSE** [reply]
> Doing this may be easier actually than trying to modify the MaterialVPP... 

Ah, so make a new VPP that contains all qp coords and merge that data with the Material VPP after-the-fact. Element and qp ids should match (right?) so that sounds reasonable.

> to get a list of ids use this: getMesh() on the MooseMesh (_mesh in the class probably) to get the libmesh meshbase then this routine to get a range of element on the blocks desired 

But if I still want _all_ elements, then I'd need to do an outer loop over all blocks? If so how do I get at all block ids? I'm assuming there's a method in that class for that...

**Reply by GiudGiud** [reply, marked as answer]
> But if I still want all elements, then I'd need to do an outer loop over all blocks? If so how do I get at all block ids? I'm assuming there's a method in that class for that...

there is also a method that does all elements. There's both! MeshBase has all the methods. 

But yeah, if you do a new class you could avoid these considerations by inheriting the right VPP class, I think

**Comment by jessecarterMOOSE** [comment]
Added #25130 based on this discussion.
